test/TritonGPU/fence-inserstion.mlir

```diff
@@ -18,3 +16,31 @@ module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-c
     tt.return
   }
 }
+
+// -----
+
+#blocked = #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [2, 16], warpsPerCTA = [8, 1], order = [1, 0]}>
+#blocked2 = #triton_gpu.blocked<{sizePerThread = [8, 1], threadsPerWarp = [16, 2], warpsPerCTA = [1, 8], order = [0, 1]}>
+#mma = #triton_gpu.nvidia_mma<{versionMajor = 3, versionMinor = 0, warpsPerCTA = [8, 1], instrShape = [16, 64, 16]}>
+#shared = #triton_gpu.shared<{vec = 8, perPhase = 1, maxPhase = 8, order = [1, 0], hasLeadingOffset = true}>
+#shared1 = #triton_gpu.shared<{vec = 8, perPhase = 1, maxPhase = 8, order = [0, 1], hasLeadingOffset = true}>
+module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 32 : i32} {
+  // CHECK-LABEL: fence_outside_loop
+  tt.func public @fence_outside_loop(%arg0: tensor<128x128xf16, #blocked>, %arg1: tensor<128x64xf16, #shared1>) {
+    %cst = arith.constant dense<0.000000e+00> : tensor<128x64xf32, #mma>
+    %c64_i32 = arith.constant 64 : i32
+    %c0_i32 = arith.constant 0 : i32
+    %c32_i32 = arith.constant 32 : i32
+    %0 = triton_gpu.convert_layout %arg0 : tensor<128x128xf16, #blocked> -> tensor<128x128xf16, #shared>
+    // CHECK: triton_nvidia_gpu.fence_async_shared
+    // CHECK: scf.for
+    // CHECK-NOT: triton_nvidia_gpu.fence_async_shared
+    // CHECK:   tt.dot
+    scf.for %iv0 = %c0_i32 to %c64_i32 step %c32_i32 : i32 {
+      scf.for %iv1 = %c0_i32 to %c64_i32 step %c32_i32 : i32 {
+        %2 = tt.dot %0, %arg1, %cst {allowTF32 = true, maxNumImpreciseAcc = 0 : i32} : tensor<128x128xf16, #shared> * tensor<128x64xf16, #shared1> -> tensor<128x64xf32, #mma>
+      }
+    }
+    tt.return
+  }
+}
```

Comment:
If it's important that we don't hoist when there are two operands with shared dependencies, should we write a testcase for that?

Reply:
well I only handle one operand for simplicity, not sure it is worth the extra lit here.

Reply:
Yeah, if it's not a correctness issue then it's fine.